File: src/fable/Fable.Core/ts/Date.ts
import { create as timeSpanCreate } from "./TimeSpan"
import { compare as utilCompare } from "./Util"
import * as Long from "./Long"

export const enum DateKind {
  UTC = 1,
  Local = 2
}

function __changeKind(d: Date, kind: DateKind) {
  let d2: Date;
  return (<any>d).kind == kind ? d : (d2 = new Date(d.getTime()), (<any>d2).kind = kind, d2);
}

function __getValue(d: Date, key: string): number {
  return (<any>d)[((<any>d).kind == DateKind.UTC ? "getUTC" : "get") + key]();
}

export function minValue() {
  return parse(-8640000000000000, 1);
}

export function maxValue() {
  return parse(8640000000000000, 1);
}

export function parse(v?: any, kind?: DateKind): any {
  const date = (v == null) ? new Date() : new Date(v);
  if (isNaN(date.getTime()))
    throw new Error("The string is not a valid Date.");
  (<any>date).kind = kind ||
    (typeof v == "string" && v.slice(-1) == "Z" ? DateKind.UTC : DateKind.Local);
  return date;
}

export function tryParse(v: any): [boolean, Date] {
  try {
    return [true, parse(v)];
  }
  catch (_err) {
    return [false, minValue()];
  }
}

export function create(year: number, month: number, day: number, h: number = 0, m: number = 0, s: number = 0, ms: number = 0, kind: DateKind = DateKind.Local) {
  const date: Date = (kind === DateKind.UTC)
    ? new Date(Date.UTC(year, month - 1, day, h, m, s, ms))
    : new Date(year, month - 1, day, h, m, s, ms);
  if (isNaN(date.getTime()))
    throw new Error("The parameters describe an unrepresentable Date.");
  (<any>date).kind = kind;
  return date;
}

export function now() {
  return parse();
}

export function utcNow() {
  return parse(null, 1);
}

export function today() {
  return date(now());
}

export function isLeapYear(year: number) {
  return year % 4 == 0 && year % 100 != 0 || year % 400 == 0;
}

export function daysInMonth(year: number, month: number) {
  return month == 2
    ? isLeapYear(year) ? 29 : 28
    : month >= 8 ? month % 2 == 0 ? 31 : 30 : month % 2 == 0 ? 30 : 31;
}

export function toUniversalTime(d: Date) {
  return __changeKind(d, 1);
}

export function toLocalTime(d: Date) {
  return __changeKind(d, 2);
}

export function timeOfDay(d: Date) {
  return timeSpanCreate(0, hour(d), minute(d), second(d), millisecond(d));
}

export function date(d: Date) {
  return create(year(d), month(d), day(d), 0, 0, 0, 0, (<any>d).kind);
}

export function day(d: Date) {
  return __getValue(d, "Date");
}

export function hour(d: Date) {
  return __getValue(d, "Hours");
}

export function millisecond(d: Date) {
  return __getValue(d, "Milliseconds");
}

export function minute(d: Date) {
  return __getValue(d, "Minutes");
}

export function month(d: Date) {
  return __getValue(d, "Month") + 1;
}

export function second(d: Date) {
  return __getValue(d, "Seconds");
}

export function year(d: Date) {
  return __getValue(d, "FullYear");
}

export function ticks(d: Date) {
  return Long.fromNumber(d.getTime())
             .add(62135596800000) // UnixEpochMilliseconds
             .sub((<any>d).kind == DateKind.Local ? d.getTimezoneOffset()*60*1000 : 0)
             .mul(10000);
}

export function toBinary(d: Date) {
  return ticks(d);
}

export function dayOfWeek(d: Date) {
  return __getValue(d, "Day");
}

export function dayOfYear(d: Date) {
  const _year = year(d);
  const _month = month(d);
  let _day = day(d);
  for (let i = 1; i < _month; i++)
    _day += daysInMonth(_year, i);
  return _day;
}

export function add(d: Date, ts: number) {
  return parse(d.getTime() + <number>ts, (<any>d).kind);
}

export function addDays(d: Date, v: number) {
  return parse(d.getTime() + v * 86400000, (<any>d).kind);
}

export function addHours(d: Date, v: number) {
  return parse(d.getTime() + v * 3600000, (<any>d).kind);
}

export function addMinutes(d: Date, v: number) {
  return parse(d.getTime() + v * 60000, (<any>d).kind);
}

export function addSeconds(d: Date, v: number) {
  return parse(d.getTime() + v * 1000, (<any>d).kind);
}

export function addMilliseconds(d: Date, v: number) {
  return parse(d.getTime() + v, (<any>d).kind);
}

export function addTicks(d: Date, t: Long.Long) {
  return parse(Long.fromNumber(d.getTime()).add(t.div(10000)).toNumber(), (<any>d).kind);
}

export function addYears(d: Date, v: number) {
  const newMonth = month(d);
  const newYear = year(d) + v;
  const _daysInMonth = daysInMonth(newYear, newMonth);
  const newDay = Math.min(_daysInMonth, day(d));
  return create(newYear, newMonth, newDay, hour(d), minute(d), second(d), millisecond(d), (<any>d).kind);
}

export function addMonths(d: Date, v: number) {
  let newMonth = month(d) + v;
  let newMonth_ = 0;
  let yearOffset = 0;
  if (newMonth > 12) {
    newMonth_ = newMonth % 12;
    yearOffset = Math.floor(newMonth / 12);
    newMonth = newMonth_;
  } else if (newMonth < 1) {
    newMonth_ = 12 + newMonth % 12;
    yearOffset = Math.floor(newMonth / 12) + (newMonth_ == 12 ? -1 : 0);
    newMonth = newMonth_;
  }
  const newYear = year(d) + yearOffset;
  const _daysInMonth = daysInMonth(newYear, newMonth);
  const newDay = Math.min(_daysInMonth, day(d));
  return create(newYear, newMonth, newDay, hour(d), minute(d), second(d), millisecond(d), (<any>d).kind);
}

export function subtract(d: Date, that: Date | number) {
  return typeof that == "number"
    ? parse(d.getTime() - <number>that, (<any>d).kind)
    : d.getTime() - (<Date>that).getTime();
}

export function toLongDateString(d: Date) {
  return d.toDateString();
}

export function toShortDateString(d: Date) {
  return d.toLocaleDateString();
}

export function toLongTimeString(d: Date) {
  return d.toLocaleTimeString();
}

export function toShortTimeString(d: Date) {
  return d.toLocaleTimeString().replace(/:\d\d(?!:)/, "");
}

export function equals(d1: Date, d2: Date) {
  return d1.getTime() == d2.getTime();
}

export function compare(x: Date, y: Date) {
  return utilCompare(x, y)
}

export function compareTo(x: Date, y: Date) {
  return utilCompare(x, y)
}

export function op_Addition(x: Date, y: number) {
  return add(x, y);
}

export function op_Subtraction(x: Date, y: number | Date) {
  return subtract(x, y);
}

File: src/fable-library-dart/Date.dart
// ignore_for_file: file_names, constant_identifier_names
import 'dart:math' as math;
import 'TimeSpan.dart' as timespan;
import 'Types.dart' as types;
import 'Util.dart' as util;

// We define Unspecified for .NET compatibility but it's not used in Dart DateTime
class DateTimeKind {
  static const Unspecified = 0;
  static const Utc = 1;
  static const Local = 2;
}

// Dart DateTime limits, from https://stackoverflow.com/a/67148809
// Use .NET values instead? Problem is we cannot set DateTimeKind.Unspecified

DateTime minValue() => DateTime.utc(-271821, 04, 20);
DateTime maxValue() => DateTime.utc(275760, 09, 13);

const _epochMicrosecondsOffset = 62135596800000000;

int unixEpochMicrosecondsToTicks(int ms) {
  return (ms + _epochMicrosecondsOffset) * 10;
}

int ticksToUnixEpochMicroseconds(int ticks) {
  return ticks ~/ 10 - _epochMicrosecondsOffset;
}

void _assertNonUnspecified(int kind) {
  if (kind == DateTimeKind.Unspecified) {
    throw Exception("DateTimeKind.Unspecified is not supported");
  }
}

final _parseFormat = RegExp(
    r'^\+?(?<date>(?<date1>\d+)[\/\-.](?<date2>\d+)(?:[\/\-.](?<date3>\d+))?(?:T|\s+))?(?<time>\d+:\d+(?::\d+(?:\.\d+)?)?)?\s*(?<ampm>[AaPp][Mm])?\s*(?<offset>Z|[+-](?<offsethours>[01]?\d):?(?<offsetminutes>[0-5]?\d)?)?$');

DateTime parse(String input) {
  input = input.trim();

  final m = _parseFormat.firstMatch(input);
  if (m == null) {
    throw Exception("Invalid date format: $input");
  }

  var utc = false;
  int offsetTotalMinutes = 0;
  int year = 0;
  int month = 0;
  int day = 0;
  int hours = 0;
  int minutes = 0;
  double seconds = 0;
  int microSeconds = 0;

  // Assume MDY by default
  final datePart = m.namedGroup("date");
  if (datePart != null) {
    final date3_ = m.namedGroup("date3");
    month = int.parse(m.namedGroup("date1")!);
    day = int.parse(m.namedGroup("date2")!);
    year = date3_ != null ? int.parse(date3_) : DateTime.now().year;
    // YMD
    if (month > 31) {
      final tmp = month;
      month = day;
      day = year;
      year = tmp;
    }
    // .NET Invariant Culture doesn't accept DMY
    // else if (month > 12) {
    //   final tmp = month;
    //   month = day;
    //   day = tmp;
    // }
  } else {
    var d = DateTime.now();
    year = d.year;
    month = d.month;
    day = d.day;
  }

  final time = m.namedGroup("time");
  if (time != null) {
    final timeParts = time.split(":");
    hours = int.parse(timeParts[0]);
    if (timeParts.length > 1) {
      minutes = int.parse(timeParts[1]);
    }
    if (timeParts.length > 2) {
      seconds = double.parse(timeParts[2]);
    }
    final amPm = m.namedGroup("ampm");
    if (amPm != null && amPm.toLowerCase() == 'pm') {
      hours += 12;
    }
  }

  int secondsInt = seconds.floor();
  if (seconds > secondsInt) {
    microSeconds = ((seconds - secondsInt) * 1000000).toInt();
  }

  final offset = m.namedGroup("offset");
  if (offset != null) {
    utc = true;
    if (offset.toUpperCase() != "Z") {
      final offsetMinutes = m.namedGroup("offsetminutes");
      offsetTotalMinutes = int.parse(m.namedGroup("offsethours")!) * 60 +
          (offsetMinutes != null ? int.parse(offsetMinutes) : 0);
      if (offset.startsWith("-")) {
        offsetTotalMinutes = -offsetTotalMinutes;
      }
    }
  }

  final parsedDate = utc
      ? DateTime.utc(
          year, month, day, hours, minutes, secondsInt, 0, microSeconds)
      : DateTime(year, month, day, hours, minutes, secondsInt, 0, microSeconds);

  // DateTime will be UTC in this case so it's safe to add time, see:
  // https://medium.com/pinch-nl/datetime-and-daylight-saving-time-in-dart-9c9468633b5d
  if (offsetTotalMinutes != 0) {
    parsedDate.add(Duration(minutes: offsetTotalMinutes));
  }
  // .NET always parses as local, even if input is UTC
  return parsedDate.isUtc ? parsedDate.toLocal() : parsedDate;
}

bool tryParse(String input, types.FSharpRef<DateTime> defaultValue) {
  try {
    defaultValue.contents = parse(input);
    return true;
  } catch (_) {
    return false;
  }
}

DateTime create(int year, int month, int day,
    [int hour = 0,
    int min = 0,
    int sec = 0,
    int millisec = 0,
    int kind = DateTimeKind.Local]) {
  _assertNonUnspecified(kind);
  return kind == DateTimeKind.Utc
      ? DateTime.utc(year, month, day, hour, min, sec, millisec)
      : DateTime(year, month, day, hour, min, sec, millisec);
}

DateTime specifyKind(DateTime d, int kind) {
  _assertNonUnspecified(kind);
  return kind == DateTimeKind.Utc ? d.toUtc() : d.toLocal();
}

Duration timeOfDay(DateTime d) {
  return Duration(
      hours: d.hour,
      minutes: d.minute,
      seconds: d.second,
      milliseconds: d.millisecond,
      microseconds: d.microsecond);
}

DateTime date(DateTime d) {
  return d.isUtc
      ? DateTime.utc(d.year, d.month, d.day)
      : DateTime(d.year, d.month, d.day);
}

int year(DateTime d) => d.year;
int month(DateTime d) => d.month;
int day(DateTime d) => d.day;
int hour(DateTime d) => d.hour;
int minute(DateTime d) => d.minute;
int millisecond(DateTime d) => d.millisecond;
int dayOfWeek(DateTime d) => d.weekday % 7;

int dayOfYear(DateTime d) {
  final _year = d.year;
  final _month = d.month;
  var _day = d.day;
  for (var i = 1; i < _month; i++) {
    _day += daysInMonth(_year, i);
  }
  return _day;
}

// Don't use Dart Duration because it can cause issues with daylight saving time, see
// https://www.flutterclutter.dev/flutter/troubleshooting/datetime-add-and-subtract-daylight-saving-time/2021/2317/

DateTime addYears(DateTime d, int v) {
  final newMonth = d.month;
  final newYear = d.year + v;
  final _daysInMonth = daysInMonth(newYear, newMonth);
  final newDay = math.min(_daysInMonth, d.day);
  return d.isUtc
      ? DateTime.utc(newYear, newMonth, newDay, d.hour, d.minute, d.second,
          d.millisecond, d.microsecond)
      : DateTime(newYear, newMonth, newDay, d.hour, d.minute, d.second,
          d.millisecond, d.microsecond);
}

DateTime addMonths(DateTime d, int v) {
  var newMonth = d.month + v;
  var newMonth_ = 0;
  var yearOffset = 0;
  if (newMonth > 12) {
    newMonth_ = newMonth.remainder(12);
    // If we use ~/ operator here, it doesn't work with negative numbers
    // See difference between truncate (what ~/ does) and floor
    yearOffset = (newMonth / 12).floor();
    newMonth = newMonth_;
  } else if (newMonth < 1) {
    newMonth_ = 12 + newMonth.remainder(12);
    yearOffset = (newMonth / 12).floor() + (newMonth_ == 12 ? -1 : 0);
    newMonth = newMonth_;
  }
  final newYear = d.year + yearOffset;
  final _daysInMonth = daysInMonth(newYear, newMonth);
  final newDay = math.min(_daysInMonth, d.day);
  return d.isUtc
      ? DateTime.utc(newYear, newMonth, newDay, d.hour, d.minute, d.second,
          d.millisecond, d.microsecond)
      : DateTime(newYear, newMonth, newDay, d.hour, d.minute, d.second,
          d.millisecond, d.microsecond);
}

DateTime add(DateTime d, Duration ts) {
  final newDate = d.add(ts);
  if (!d.isUtc) {
    final oldTzOffset = d.timeZoneOffset;
    final newTzOffset = newDate.timeZoneOffset;
    return oldTzOffset != newTzOffset
        // Maxime Mangel: I am not sure why but the next line needs to have
        // different reversed compared to other languages implementations
        // Without that, I have an error on the test "Adding days to a local date works even if daylight saving time changes"
        // when executing from Europe/Paris timezone
        // If this breaks another timezone, we need to investigate for a more robust detection
        ? newDate.add(oldTzOffset - newTzOffset)
        : newDate;
  } else {
    return newDate;
  }
}

DateTime subtract(DateTime d, Duration ts) {
  return add(d, -ts);
}

Duration subtractDate(DateTime d1, DateTime d2) {
  return Duration(
      microseconds: d1.microsecondsSinceEpoch - d2.microsecondsSinceEpoch);
}

int compare(DateTime dt1, DateTime dt2) {
  return dt1.compareTo(dt2);
}

DateTime addDays(DateTime d, double v) => add(d, timespan.fromDays(v));
DateTime addHours(DateTime d, double v) => add(d, timespan.fromHours(v));
DateTime addMinutes(DateTime d, double v) => add(d, timespan.fromMinutes(v));
DateTime addSeconds(DateTime d, double v) => add(d, timespan.fromSeconds(v));
DateTime addMilliseconds(DateTime d, double v) =>
    add(d, timespan.fromMilliseconds(v));
DateTime addTicks(DateTime d, int v) => add(d, Duration(microseconds: v ~/ 10));

DateTime now() {
  return DateTime.now();
}

DateTime utcNow() {
  return DateTime.now().toUtc();
}

DateTime today() {
  return date(DateTime.now());
}

bool isLeapYear(int year) =>
    year % 4 == 0 && year % 100 != 0 || year % 400 == 0;

int daysInMonth(int year, int month) => month == 2
    ? (isLeapYear(year) ? 29 : 28)
    : (month >= 8 ? (month % 2 == 0 ? 31 : 30) : (month % 2 == 0 ? 30 : 31));

DateTime fromTicks(int ticks, [int kind = DateTimeKind.Local]) {
  _assertNonUnspecified(kind);
  final ms = ticksToUnixEpochMicroseconds(ticks);
  return DateTime.fromMicrosecondsSinceEpoch(ms,
      isUtc: kind == DateTimeKind.Utc);
}

int getTicks(DateTime date) {
  return unixEpochMicrosecondsToTicks(date.microsecondsSinceEpoch);
}

int kind(DateTime dt) => dt.isUtc ? DateTimeKind.Utc : DateTimeKind.Local;

String dateOffsetToString(Duration offset) {
  var offsetMinutes = offset.inMinutes;
  final isMinus = offsetMinutes < 0;
  offsetMinutes = offsetMinutes.abs();
  final hours = offsetMinutes ~/ 3600000;
  final minutes = (offsetMinutes % 3600000) ~/ 60000;
  return (isMinus ? "-" : "+") +
      util.padWithZeros(hours, 2) +
      ":" +
      util.padWithZeros(minutes, 2);
}

String toDateOnlyString(DateTime date) {
  final str = date.toIso8601String();
  return str.substring(0, str.indexOf("T"));
}

String toTimeOnlyString(DateTime date) {
  final str = date.toIso8601String();
  return str.substring(str.indexOf("T") + 1, str.length - (date.isUtc ? 1 : 0));
}

String dateToISOString(DateTime d, [bool? utc]) {
  if (utc == null) {
    utc = d.isUtc;
  } else {
    d = utc ? d.toUtc() : d.toLocal();
  }
  return utc
      ? d.toIso8601String()
      : d.toIso8601String() + dateOffsetToString(d.timeZoneOffset * -60000);
}

String dateToStringWithCustomFormat(DateTime date, String format, [bool? utc]) {
  if (utc != null) {
    date = utc ? date.toUtc() : date.toLocal();
  }
  return format.replaceAllMapped(RegExp(r'(\w)\1*'), (Match match) {
    int? rep;
    final matchValue = match.group(0)!;
    switch (matchValue.substring(0, 1)) {
      case "y":
        final y = date.year;
        rep = matchValue.length < 4 ? y % 100 : y;
        break;
      case "M":
        rep = date.month;
        break;
      case "d":
        rep = date.day;
        break;
      case "H":
        rep = date.hour;
        break;
      case "h":
        final h = date.hour;
        rep = h > 12 ? h % 12 : h;
        break;
      case "m":
        rep = date.minute;
        break;
      case "s":
        rep = date.second;
        break;
      case "f":
        rep = date.millisecond;
        break;
    }
    if (rep == null) {
      return matchValue;
    } else {
      return (rep < 10 && matchValue.length > 1)
          ? "0" + rep.toString()
          : "" + rep.toString();
    }
  });
}

String toString(DateTime date, [String? format]) {
  if (format == null) {
    return date.toString();
  } else if (format.length == 1) {
    switch (format) {
      case "D":
      case "d":
        return toDateOnlyString(date);
      case "T":
      case "t":
        return toTimeOnlyString(date);
      case "O":
      case "o":
        return dateToISOString(date);
      default:
        throw Exception("Unrecognized Date print format");
    }
  } else {
    return dateToStringWithCustomFormat(date, format);
  }
}

// Doesn't seem to work
// bool _isDST(Duration janOffset, Duration julOffset, Duration tOffset) {
//   final min = janOffset > julOffset ? julOffset : janOffset;
//   return min == tOffset;
// }

// bool isDaylightSavingTime(DateTime d) {
//   final year = d.year;
//   final jan = DateTime(year, 0, 1);
//   final jul = DateTime(year, 6, 1);
//   return _isDST(jan.timeZoneOffset, jul.timeZoneOffset, d.timeZoneOffset);
// }
